### PR TITLE
[dev-launcher] load updates by manifestPermalink

### PR DESCRIPTION
**Pull request by ajsmth**
# Why

This PR adds the ability for the dev launcher to load EAS Updates via the manifest permalink - an extension of the work @esamelson landed recently 

# How

Passed the manifest permalink to dev launchers `loadApp` method
Accounted for various loading and error states involved with loading an update
Refactored some components 

# Test Plan

I've been using the project https://u.expo.dev/6899a588-e224-4cc5-a91f-75122be68353 as a testing point, however any project configured with EAS updates should work 

Navigate to the updates page for a branch and select an individual update, ensure it launches
Navigate to the updates page for a branch and select "Open Latest", ensure it launches
Pass an update url to the url dropdown - this one works as a sample: https://u.expo.dev/update/7c1d2bd0-f88b-454d-998c-7fa92a924dbf

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-dev-launcher/bundle/functions/getRecentlyRuntime.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { Branch } from '../queries/useBranchesForApp';
+
+export function getRecentRuntime(branches: Branch[]) {
```
Comment (esamelson) [inline packages/expo-dev-launcher/bundle/functions/getRecentlyRuntime.ts:3]: probably want to rename the file to match this export name

**Review by esamelson — COMMENTED** [review]
We'll also want to make sure that the `projectURL` parameter in the native code is set correctly. When opening an EAS Update via this new UI, it should be set to the value of EXUpdatesURL, but when opening an arbitrary URL via the dropdown, it should probably be set to that URL instead. We may need to add a parameter to the exported `loadApp` method to account for this.

**Comment by ajsmth** [comment]
@esamelson I've updated the implementation - let me know what you think and I'll go ahead and do the same on Android 

I chose to allow the project url to be specified as an argument in cases where we might want to test specific apps / projects that aren't necessarily configured in the `Expo.plist` - e.g I'm envisioning a debug option that allows us to do this in the settings menu at some point in the future

File: packages/expo-dev-launcher/ios/EXDevLauncherController.m [hunk 2]
```diff
@@ -294,13 +294,36 @@ - (NSURL *)sourceUrl
   return _sourceUrl;
 }
 
-- (void)loadApp:(NSURL *)expoUrl onSuccess:(void (^ _Nullable)(void))onSuccess onError:(void (^ _Nullable)(NSError *error))onError
+- (BOOL)isUpdatesUrl:(NSURL *)url
```
Comment (esamelson) [inline packages/expo-dev-launcher/ios/EXDevLauncherController.m:297]:
```suggestion
- (BOOL)isEASUpdateURL:(NSURL *)url
```

**Review by esamelson — APPROVED** [review]
Cool, yeah, I think this looks good. You might need to pull in this PR temporarily in order to test on Android -- https://github.com/expo/expo/pull/17069

**Pull request merged**